sklearn/datasets/samples_generator.py

```diff
@@ -28,8 +27,7 @@ def _generate_hypercube(samples, dimensions, rng):
     if dimensions > 30:
         return np.hstack([_generate_hypercube(samples, dimensions - 30, rng),
                           _generate_hypercube(samples, 30, rng)])
-    out = astype(sample_without_replacement(2 ** dimensions, samples,
-                                            random_state=rng),
+    out = astype(random_state.permutation(2 ** dimensions)[:samples],
```

Comment:
go back to calling ``sample_without_replacement`` and then use the heuristic I gave to decide whether to use permutation. thanks :)